app/Platform/Services/GameTopAchieversService.php

```diff
@@ -0,0 +1,150 @@
+    {
+        return $this->masteryQuery()->with('user')
+            ->orderByDesc('last_unlock_hardcore_at')
+            ->limit($count)
+            ->get();
+    }
+
+    /**
+     * @return Collection<int, PlayerGame>
+     */
+    public function allMasteries(int $offset = 0, int $count = 10): Collection
+    {
+        return $this->masteryQuery()->with('user')
+            ->orderBy('last_unlock_hardcore_at')
+            ->offset($offset)
+            ->limit($count)
+            ->get();
+    }
+
+    /**
+     * @return Collection<int, PlayerGame>
+     */
+    public function highestPointEarners(int $count = 10): Collection
+    {
+        $query = $this->baseQuery();
+
+        if ($this->masteryPoints === 0) {
+            // event with no points. primary sort by number of achievements unlocked.
+            $query = $query->where('achievements_unlocked_hardcore', '>', 0)
+                ->orderByDesc('achievements_unlocked_hardcore');
+        } else {
+            // standard game. primary sort by number of points earned.
+            $query = $query->orderByDesc('points_hardcore');
+        }
+
+        return $query
+            ->orderBy('last_unlock_hardcore_at')
+            ->with('user')
+            ->limit($count)
+            ->get();
+    }
+
+    public static function expireTopAchieversComponentData(int $gameId): void
+    {
+        $cacheKey = "game:{$gameId}:top-achievers";
+        Cache::forget($cacheKey);
+    }
+
+    public function getTopAchieversComponentData(): array
+    {
+        $cacheKey = "game:{$this->gameId}:top-achievers";
+        $retval = Cache::get($cacheKey);
+        if ($retval !== null) {
+            return $retval;
+        }
+
+        $numMasteries = $this->numMasteries();
+        if ($numMasteries < 10) {
+            return [
+                $numMasteries,
+                $this->convertPlayerGames($this->highestPointEarners()),
+            ];
+        }
+
+        $retval = [
+            $numMasteries,
+            $this->convertPlayerGames($this->recentMasteries()),
+        ];
+
+        // only cache the result if the masters list is full.
+        // that way we only have to expire it when there's a new mastery
+        // or an achievement gets promoted or demoted
+        Cache::put($cacheKey, $retval, Carbon::now()->addDays(30));
+
+        return $retval;
+    }
+
+    /**
+     * @param Collection<int, PlayerGame> $playerGames
+     */
+    private function convertPlayerGames(Collection $playerGames): array
+    {
+        $retval = [];
+        foreach ($playerGames as $playerGame) {
+            $retval[] = [
+                'user_id' => $playerGame->user_id,
+                'achievements_unlocked_hardcore' => $playerGame->achievements_unlocked_hardcore,
+                'points_hardcore' => $playerGame->points_hardcore,
+                'last_unlock_hardcore_at' => $playerGame->last_unlock_hardcore_at->unix(),
```

Comment:
I seem to be getting an error on this line when I open a game that has only softcore unlocks and no hardcore unlocks. This does seem to imply the UI needs to support some kind of empty state (or just 404 the dedicated page and don't render the dedicated component on the game page).

![Screenshot 2024-08-02 at 5 35 48 PM](https://github.com/user-attachments/assets/11d849d2-2bcd-4671-a89f-5f59c4fcda45)

Reply:
This also happens if there aren't enough hardcore players to fill the list. I've fixed the error and added a default message if the list is empty.